runtime/compiler/optimizer/HandleRecompilationOps.cpp

```diff
@@ -0,0 +1,258 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include <stdint.h>
+#include "env/FrontEnd.hpp"
+#include "env/j9method.h"
+#include "il/Block.hpp"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/TreeTop.hpp"
+#include "il/TreeTop_inlines.hpp"
+#include "infra/Cfg.hpp"
+#include "infra/CfgEdge.hpp"
+#include "optimizer/HandleRecompilationOps.hpp"
+#include "optimizer/Optimization.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+
+int32_t
+TR_HandleRecompilationOps::perform()
+   {
+   _enableTransform = true;
```

Comment:
Why not just return 0 from the cases checked when we know we don't want to do this rather than having the global variable?

Reply:
The way the changes are set up right now, `Walker.cpp` is always generating IL for these situations rather than aborting the compilation.  It's now relying on this ILGen opt. pass to check whether inducing OSR is needed, and this opt. aborts if it cannot induce OSR.

If you feel that's the wrong approach, and that `Walker.cpp` itself should abort the compilation if OSR can't be induced, I'll switch to that way of handling it.

Reply:
Here's what I'm thinking of doing.  The conditions that are checked in `TR_HandleRecompilationOps::perform()` will also be checked during ILGen in `Walker.cpp`.  If any of those conditions is not satisfied, compilation will be aborted during ILGen itself.  That way, if `TR_HandleRecompilationOps::perform` encounters any of those conditions, it can immediately return without performing any analysis.

Some additional conditions are checked in [`OMR::ResolvedMethodSymbol::induceOSRAfterAndRecompile`](https://github.com/hzongaro/omr/blob/e2fac34fcb3571cdf5351693a04fc2852b18f46f/compiler/il/OMRResolvedMethodSymbol.cpp#L478), which is called by [`TR_HandleRecompilationOps::visitNode`](https://github.com/eclipse/openj9/blob/5a773894edd1f76514c741eb425971d9c6cd3065/runtime/compiler/optimizer/HandleRecompilationOps.cpp#L219).  The result of that call to `OMR::ResolvedMethodSymbol::induceOSRAfterAndRecompile` is currently checked with a `TR_ASSERT_FATAL`, but if the conditions that `induceOSRAfterAndRecompile` is checking could not be checked during ILGen itself, then `TR_HandleRecompilationOps::visitNode` should simply abort the compilation at that point rather than performing a fatal assertion. . . .

Reply:
Made these changes in commit 3ecec562fd3b56fad1eeef986b6ca9d71d4aee17